### PR TITLE
refactor(lua): call `loadfile` internally instead of `luaL_loadfile`

### PR DESCRIPTION
**Pull request by lewis6991**
~A version of loadfile with the intention of being redefined to allow third-party plugins to provide instrumentation and caching.~

This will allow [impatient.nvim](https://github.com/lewis6991/impatient.nvim) to cache Lua files in directories like `plugin` and `after`.

`loadfile` is still called via C so the only additional overhead is via the Lua-C API which shouldn't be too bad.

Obviously this will need some discussion (hence this is a draft), but I hope we agree on the overall goal.

EDIT: Changed so the Lua built-in `loadfile` is called instead of the direct C version `luaL_loadfile`.

### CODE REVIEW

**Comment by lewis6991** [comment]
Before I work on this any more, @bfredl are you OK with the premise of this PR?

**Comment by bfredl** [comment]
@lewis6991 yes.

**Comment by bfredl** [comment]
Needs an update to tests:

```
2022-02-02T18:10:27.5497661Z test/helpers.lua:73: Expected objects to be the same.
2022-02-02T18:10:27.5497930Z Passed in:
2022-02-02T18:10:27.5498383Z (string) 'Vim(luafile):E5113: Error while calling lua chunk: attempt to call a nil value
2022-02-02T18:10:27.5498725Z stack traceback:'
2022-02-02T18:10:27.5498958Z Expected:
2022-02-02T18:10:27.5499559Z (string) 'Vim(luafile):E5112: Error while creating lua chunk: Xtest-functional-lua-commands-luafile:1: unexpected symbol near ')''
```

I haven't checked the test in detail but likely just putting in the new message should be fine.

File: src/nvim/lua/executor.c [hunk 1]
```diff
@@ -1249,11 +1249,25 @@ bool nlua_exec_file(const char *path)
...
+
+  lua_call(lstate, 1, 2);
+
+  // loadstring() returns either:
+  //  1. nil, error
+  //  2. chunk, nil
+
+  if (lua_isnil(lstate, -2)) {
+    // 1
     nlua_error(lstate, _("E5112: Error while creating lua chunk: %.*s"));
```
Comment (bfredl) [inline src/nvim/lua/executor.c:1263]: don't we need a  `lua_pop(lstate, 1);` before return to get rid of the `nil` on stack?

Comment (lewis6991) [reply src/nvim/lua/executor.c:1263]: I don't think so, but not 100% sure. I blindly assume the stack sorts itself out in the error cases. There are many areas where we don't clear the stack, e.g: `nlua_expand_pat`, `nlua_state_init`, `nlua_call_user_expand_func`.

Comment (bfredl) [reply src/nvim/lua/executor.c:1263]: "clearing" the stack is never done, ever. This function might be called from the _global_ lua stack so it must be balanced, regardless if there is an error (which is a _vim_ error, not a lua state error) or not.

File: src/nvim/lua/executor.c [hunk 2]
```diff
@@ -1249,11 +1249,27 @@ bool nlua_exec_file(const char *path)
 {
   lua_State *const lstate = global_lstate;
 
-  if (luaL_loadfile(lstate, path)) {
+  lua_getglobal(lstate, "loadfile");
+  lua_pushstring(lstate, path);
+
+  lua_call(lstate, 1, 2);
```
Comment (bfredl) [inline src/nvim/lua/executor.c:1255]: let's use `lua_pcall` to be sure. A user override might cause an uncaught exception. (we can use the same error format E5112 for this)

Comment (lewis6991) [reply src/nvim/lua/executor.c]: Used the slightly more generic `E5111`.

**Pull request merged**